core/src/main/java/org/infinispan/expiration/impl/ClusterExpirationManager.java

```diff
@@ -223,67 +233,101 @@ private boolean pollForCompletion(BlockingQueue<CompletableFuture<?>> queue, lon
-      Object expiringObject = expiring.putIfAbsent(key, completableFuture);
-      if (expiringObject == null) {
+      CompletableFuture<Boolean> previousFuture = expiring.putIfAbsent(key, completableFuture);
+      if (previousFuture == null) {
          if (trace) {
-            log.tracef("Submitting expiration removal for key %s which had maxIdle of %s", toStr(key), maxIdle);
+            log.tracef("Submitting expiration removal for key: %s which is maxIdle: %s of: %s", toStr(key), maxIdle, time);
          }
-         completableFuture.whenComplete((b, t) -> expiring.remove(key, completableFuture));
          try {
-            AdvancedCache<K, V> cacheToUse = skipLocking ? cache.withFlags(Flag.SKIP_LOCKING) : cache;
-            CompletableFuture<Boolean> future = removeMaxIdle(cacheToUse, key, value);
+            AdvancedCache<K, V> cacheToUse = cacheToUse(isWrite);
+            CompletableFuture<Boolean> future;
+            if (maxIdle) {
+               future = removeMaxIdle(cacheToUse, key, value);
+            } else {
+               future = removeLifespan(cacheToUse, key, value, time);
+            }
             return future.whenComplete((b, t) -> {
+               // We have to remove the entry from the map before setting the exception status - otherwise retry for
+               // a write at the same time could get stuck in a recursive loop
+               expiring.remove(key);
                if (t != null) {
                   completableFuture.completeExceptionally(t);
                } else {
                   completableFuture.complete(b);
                }
             });
          } catch (Throwable t) {
+            // We have to remove the entry from the map before setting the exception status - otherwise retry for
+            // a write at the same time could get stuck in a recursive loop
+            expiring.remove(key);
             completableFuture.completeExceptionally(t);
             throw t;
          }
-      } else if (expiringObject instanceof CompletableFuture) {
-         // This means there was another thread that found it had expired via max idle
-         return (CompletableFuture<Boolean>) expiringObject;
-      } else {
-         // If it wasn't a CompletableFuture we had a lifespan removal occurring so it will be removed for sure
-         return CompletableFutures.completedTrue();
       }
+      if (isWrite) {
+         if (trace) {
+            log.tracef("Waiting on prior expiration removal for key %s as this command is a write", key);
+         }
+         return previousFuture.handle((expired, t) -> {
+            if (t != null) {
+               boolean foundTimeout = false;
+               Throwable cause = t;
+               while (cause != null) {
+                  if (cause instanceof org.infinispan.util.concurrent.TimeoutException) {
+                     foundTimeout = true;
+                     break;
+                  }
+                  cause = cause.getCause();
```

Comment:
Please don't use while loops to look for causes... `CompletableFutures.extractException(t)` should be enough.

Reply:
I can test it out with the single unwrapping.

Reply:
sadly it isn't quite enough as we can get it wrapped by a `RemoteException`, so I will add that as well. Starting to become a bit uglier again.

Reply:
Yeah, that's fine, it was just the while loop giving me a shaky ground vibe, like how many exception wrappers could there be?

I still think we should avoid the `TimeoutException`, because while we don't send any exception stack trace remotely now, we should fix https://issues.redhat.com/browse/ISPN-11022 eventually, and then the failed `RemoveExpiredCommand` RPCs won't be that cheap. 
Perhaps a method `PessimisticLockInterceptor.visitRemoteExpiredCommand` that catches the `TimeoutException` and returns `false` instead, but let's leave that for another day.